### PR TITLE
feat: support for python 3.11

### PR DESCRIPTION
**Pull request by aryarm**
This PR finally adds support for python 3.11 to haptools!

Hopefully, we can get this PR merged before bioconda tries to test our package with 3.11 in January ([source](https://github.com/bioconda/bioconda-recipes/issues/33333#issuecomment-1865116088)).

**Note:** This PR also upgrades the versions of poetry and python that we use in our development environment. Once this is merged, haptools developers will need to create a new development environment by following the updated directions in [the contributing docs](https://haptools--207.org.readthedocs.build/en/207/project_info/contributing.html#dev-setup).

### CODE REVIEW

**Comment by aryarm** [comment]
we're currently blocked by https://github.com/cjolowicz/nox-poetry/issues/1116

Otherwise, if we can't find a solution there, we'll have to drop support for python 3.7 in order to support python 3.11

**Comment by aryarm** [comment]
@mlamkin7, do you have any advice for the failing macos check?

https://github.com/CAST-genomics/haptools/actions/runs/7304778975/job/19907440862?pr=207

it's something to do with GenotypesTR, but it only happens on macs and not linux...

**Comment by aryarm** [comment]
I almost can't believe I actually finally got all of the checks to pass lol

@mlamkin7, when you have a chance, can you review this?

Unfortunately, the approach I took isn't great. Because of the issue I mentioned in https://github.com/CAST-genomics/haptools/pull/207#issuecomment-1514241521, I wasn't able to lock our dependencies for the python 3.11 checks. This means that our CI may occasionally fail for python 3.11 because of reasons that are out of our control. For example, if one of our dependencies uploads a new release with a bug in it, our python 3.11 check will be run with that bug. If that happens, we won't have any way to deal with it except to wait until the bug gets patched.

Fortunately, I have an idea for a longer term solution: [multiple constraint dependencies](https://python-poetry.org/docs/dependency-specification/#multiple-constraints-dependencies)! We can constrain `numpy>=0.20.0` for `python==3.7` but constrain `numpy>=0.21.5` for `python>=3.8`. Unfortunately, that won't currently work because of https://github.com/python-poetry/poetry-plugin-export/issues/183. But once that issue gets resolved, we should be able to [revert this code in our Noxfile](https://github.com/CAST-genomics/haptools/pull/207/files#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835R42-R84) and update [the macos check to run on py 3.11 instead of py3.9 as it does now](https://github.com/CAST-genomics/haptools/pull/207/files#diff-1db27d93186e46d3b441ece35801b244db8ee144ff1405ca27a163bfe878957fR20).

File: pyproject.toml [hunk 1]
```diff
@@ -10,58 +14,30 @@ documentation = "https://haptools.readthedocs.io"
...
+sphinx-click = ">=3.0.2"
 
+[tool.poetry.group.tests.dependencies]
 pytest = ">=6.2.5"
 black = ">=22.3.0"
 ipython = ">=7.30.1"
-nox = ">=2022.8.7"
-nox-poetry = ">=1.0.1"
+nox = ">=2022.11.21"
+nox-poetry = ">=1.0.2"
```
Comment (aryarm) [inline pyproject.toml:37]:
```suggestion
nox-poetry = ">=1.0.3"
```

this should probably match whatever we have in the constraints file

note that the lock file will also need to be updated after this

File: .devcontainer.json [hunk 2]
```diff
@@ -29,7 +29,7 @@
 				"python.defaultInterpreterPath": "/opt/conda/envs/haptools/bin/python",
 				"python.terminal.activateEnvironment": true,
 				"python.terminal.activateEnvInCurrentTerminal": true,
-				"python.venvFolders": ["/opt/conda/envs"],
+				"python.venvFolders": ["/home/vscode/.cache/pypoetry/virtualenvs"],
```
Comment (mlamkin7) [inline .devcontainer.json:32]: For line 32 is this path its referencing correct?

**Review by mlamkin7 — COMMENTED** [review]
My only question is a comment on the .devcontainer.json file but once completed the rest looks good to me.

**Review by mlamkin7 — APPROVED** [review]

**Pull request merged**